bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/MPSBulkSynchronizer.kt

```diff
@@ -141,6 +151,118 @@ object MPSBulkSynchronizer {
+                        // Synchronizing the parent will automatically update the containment of the children.
+                    }
+
+                    override fun childrenChanged(nodeId: Long, role: String?) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun referenceChanged(nodeId: Long, role: String) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun propertyChanged(nodeId: Long, role: String) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun nodeRemoved(nodeId: Long) {
+                        // same reason as in containmentChanged
+                    }
+
+                    override fun nodeAdded(nodeId: Long) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+                })
+                val treePointer = TreePointer(newTree)
+                treePointer.runRead {
+                    ModelSynchronizer(
+                        invalidationTree,
+                        treePointer.getRootNode(),
+                        MPSRepositoryAsNode(repository),
+                        NodeAssociationFromMPS2Server(true, MPSArea(repository), treePointer)
+                    ).synchronize()
+                }
+            }
+
+            val allModules = repository.modules
+            val includedModules: Iterable<SModule> = allModules.filter {
+                isModuleIncluded(it.moduleName!!, includedModuleNames, includedModulePrefixes)
+            }
+            val numIncludedModules = includedModules.count()
+            val repoAsNode = MPSRepositoryAsNode(repository)
+            println("Importing modules...")
+            try {
+                println("Importing modules...")
+                // `modulesToImport` lazily produces modules to import
+                // so that loaded model data can be garbage collected.
+                val modulesToImport = includedModules.asSequence().flatMapIndexed { index, module ->
+                    println("Importing module ${index + 1} of $numIncludedModules: '${module.moduleName}'")
+                    val fileName = inputPath + File.separator + module.moduleName + ".json"
+                    val moduleFile = File(fileName)
+                    if (moduleFile.exists()) {
+                        val expectedData: ModelData = moduleFile.inputStream().use(Json::decodeFromStream)
+                        sequenceOf(ExistingAndExpectedNode(MPSModuleAsNode(module), expectedData))
+                    } else {
+                        println("Skip importing ${module.moduleName}} because $fileName does not exist.")
+                        sequenceOf()
+                    }
```

Comment:
How do other sync directions work in this case? Do they raise an error in case the user requested syncing something that doesn't exist or do they skip it?

CC @mhuster23

Reply:
They skip it.